cellmlmanip/units.py

```diff
@@ -281,29 +367,45 @@ def traverse(self, expr: sympy.Expr):
         """descends the Sympy expression and performs Pint unit arithmetic on sub-expressions
         :param expr: a Sympy expression
         :returns: the quantity (i.e. expression * unit) of the expression
+        :throws: KeyError - if variable not found in metadata
+                 InvalidUnitsError - if expression just cannot calculate
+                 UnexpectedMathUnitsError - if math is not supported
+                 BooleanUnitsError - if math returns booleans
+                 InputArgumentsMustBeDimensionlessError - if input arguments should be dimensionless
+                 InputArgumentsInvalidUnitsError - if input arguments should have same units
+                 InputArgumentMustBeNumberError - if one of input arguments should be a number
+        NOTE: Sympy will throw exceptions if the expression is badly formed before we even get here
         """
         # collect the units for each argument of the expression (might be sub-expression)
         quantity_per_arg = []
 
-        # If the expression is a piecewise
+        if expr.is_Matrix:
```

Comment:
Is this not caught by the generic case at the bottom?

(Would up your test coverage if it was and you could remove this, by the way, at the moment the final `else` is uncovered)

Reply:
Weird one. No, if you create a matrix expression - so expr.is_matrix is True then expr.is_Piecewise does not exist and code crashes.

I put this in but did then stop testing the other edge cases as we shouldnt encounter them from cellml files anyway.